filters/auth/oauth.go

```diff
@@ -0,0 +1,437 @@
+	return &tokeninfoSpec{typ: checkOAuthTokeninfoAnyKV, tokeninfoURL: OAuthTokeninfoURL}
+}
+
+func (s *tokeninfoSpec) Name() string {
+	switch s.typ {
+	case checkOAuthTokeninfoAnyScopes:
+		return OAuthTokeninfoAnyScopeName
+	case checkOAuthTokeninfoAllScopes:
+		return OAuthTokeninfoAllScopeName
+	case checkOAuthTokeninfoAnyKV:
+		return OAuthTokeninfoAnyKVName
+	case checkOAuthTokeninfoAllKV:
+		return OAuthTokeninfoAllKVName
+	}
+	return AuthUnknown
+}
+
+// CreateFilter creates an auth filter. All arguments have to be
+// strings. Depending on the variant of the auth filter, the arguments
+// represent scopes or key-value pairs to be checked in the tokeninfo
+// response. How scopes or key value pairs are checked is based on the
+// type. The shown example for checkOAuthTokeninfoAllScopes will grant
+// access only to tokens, that have scopes read-x and write-y:
+//
+//     s.CreateFilter(read-x", "write-y")
+//
+func (s *tokeninfoSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) == 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	ac, err := newAuthClient(s.tokeninfoURL)
+	if err != nil {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	f := &filter{typ: s.typ, authClient: ac, kv: make(map[string]string)}
+	if len(sargs) > 0 {
+		switch f.typ {
+		// all scopes
+		case checkOAuthTokeninfoAllScopes:
+			fallthrough
+		case checkOAuthTokeninfoAnyScopes:
+			f.scopes = sargs[:]
+		// key value pairs
+		case checkOAuthTokeninfoAnyKV:
+			fallthrough
+		case checkOAuthTokeninfoAllKV:
+			for i := 0; i+1 < len(sargs); i += 2 {
+				f.kv[sargs[i]] = sargs[i+1]
+			}
+			if len(sargs) == 0 || len(sargs)%2 != 0 {
+				return nil, filters.ErrInvalidFilterParameters
+			}
+		default:
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return f, nil
+}
+
+func (kv kv) String() string {
+	var res []string
+	for k, v := range kv {
+		res = append(res, k, v)
+	}
+	return strings.Join(res, ",")
+}
+
+// String prints nicely the filter configuration based on the
+// configuration and check used.
+func (f *filter) String() string {
+	switch f.typ {
+	case checkOAuthTokeninfoAnyScopes:
+		return fmt.Sprintf("%s(%s)", OAuthTokeninfoAnyScopeName, strings.Join(f.scopes, ","))
+	case checkOAuthTokeninfoAllScopes:
+		return fmt.Sprintf("%s(%s)", OAuthTokeninfoAllScopeName, strings.Join(f.scopes, ","))
+	case checkOAuthTokeninfoAnyKV:
+		return fmt.Sprintf("%s(%s)", OAuthTokeninfoAnyKVName, f.kv)
+	case checkOAuthTokeninfoAllKV:
+		return fmt.Sprintf("%s(%s)", OAuthTokeninfoAllKVName, f.kv)
+	}
+	return AuthUnknown
+}
+
+func (f *filter) validateAnyScopes(h map[string]interface{}) bool {
+	if len(f.scopes) == 0 {
+		return true
+	}
+
+	vI, ok := h[scopeKey]
+	if !ok {
+		return false
+	}
+	v, ok := vI.([]interface{})
+	if !ok {
+		return false
+	}
+	var a []string
```

Comment:
the conversion of []interface{} to []string appears both in validateAnyScopes and validateAllScopes. It can be it's own function, or the intersect and all functions can accept []interface{} as the second argument and have the conversion in place, but it's also fine like this.

Reply:
I like it more like this.